text/0000-label-break-value.md

```diff
@@ -0,0 +1,119 @@
+- Feature Name: label_break_value
+- Start Date: 2017-06-26
+- RFC PR: [#2046](https://github.com/rust-lang/rfcs/pull/2046)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow a `break` of labelled blocks with no loop, which can carry a value.
+
+# Motivation
+[motivation]: #motivation
+
+In its simplest form, this allows you to terminate a block early, the same way that `return` allows you to terminate a function early.
+
+```
+'block: {
+    do_thing();
+    if condition_not_met() {
+        break 'block;
+    }
+    do_next_thing();
+    if condition_not_met() {
+        break 'block;
+    }
+    do_last_thing();
+}
+```
+In the same manner as `return` and the labelled loop breaks in [RFC 1624](https://github.com/rust-lang/rfcs/blob/master/text/1624-loop-break-value.md), this `break` can carry a value:
+```
+let result = 'block: {
+    if foo() { break 'block 1; }
+    if bar() { break 'block 2; }
+    3
+};
+```
+RFC 1624 opted not to allow options to be returned from `for` or `while` loops, since no good option could be found for the syntax, and it was hard to do it in a natural way. This proposal gives us a natural way to handle such loops with no changes to their syntax:
+```
+let result = 'block: {
+    for &v in container.iter() {
+        if v > 0 { break 'block v; }
+    }
+    0
+};
+```
+This extension handles searches more complex than loops in the same way:
+```
+let result = 'block: {
+    for &v in first_container.iter() {
+        if v > 0 { break 'block v; }
+    }
+    for &v in second_container.iter() {
+        if v < 0 { break 'block v; }
+    }
+    0
+};
+```
+Implementing this without a labelled break is much less clear:
```

Comment:
withoutboats: That's a lot neater than the iterator based solution [I proposed](https://github.com/rust-lang/rfcs/pull/2046#issuecomment-312447041).

joshtriplett: the find condition is different for the two iterators.